### PR TITLE
[SPARK-7378] [core] Handle deep links to unloaded apps.

### PR DESCRIPTION
**Pull request by vanzin**
The code was treating deep links as if they were attempt IDs, so
for example if you tried to load "/history/app1/jobs" directly,
that would fail because the code would treat "jobs" as an attempt id.

This change modifies the code to try both cases - first without an
attempt id, then with it, so that deep links are handled correctly.
This assumes that the links in the Spark UI do not clash with the
attempt id namespace, though, which is the case for YARN at least,
which is the only backend that currently publishes attempt IDs.

### CODE REVIEW

**Comment by vanzin** [comment]
Streaming failure, sigh.

**Comment by vanzin** [comment]
Jenkins, retest this please.

**Comment by vanzin** [comment]
@andrewor14 @srowen could you guys take a look at this? This is a regression from 1.3.

File: core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala [hunk 1]
```diff
@@ -82,31 +82,27 @@ class HistoryServer(
         return
       }
 
-      val appKey =
-        if (parts.length == 3) {
-          s"${parts(1)}/${parts(2)}"
-        } else {
-          parts(1)
+      val appId = parts(1)
+      val attemptId = if (parts.length > 3) Some(parts(2)) else None
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: should this be `>= 3`?

Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: Good catch. Let me retest and update the patch.

**Comment by srowen** [comment]
LGTM

**Comment by andrewor14** [comment]
LGTM master 1.4

**Pull request closed without merging**